File: cdt.cpp
/// Causal Dynamical Triangulations in C++ using CGAL
///
/// Copyright (c) 2013 Adam Getchell
///
/// A program that generates spacetimes
///
/// Inspired by https://github.com/ucdavis/CDT
///

#include <CGAL/Timer.h>

// CDT headers
#include <utilities.h>
#include <spherical_3_complex.h>
#include <periodic_3_complex.h>

int main(int argc, char* argv[]) {
  /// Start running time
  CGAL::Timer t;
  t.start();

  int dimensions = 3;         /// Number of dimensions, defaults to 3
  int num_simplices = 0;      /// Number of simplices, defaults to 0
  char topology;              /// Topology type
  bool topology_set = false;  /// Is the Topology type set?
  int c;                      /// Case statement switch integer
  opterr = 0;                 /// Suppress getopt error messages
  bool error_flag = false;    /// Is there an error in program invocation?

  if (argc < 2) {
    print_error(argv[0]);
    return 1;
  }

  /// Use getopt to parse command line arguments
  while ((c = getopt (argc, argv, "d:s:t:")) != -1)
    switch (c) {
        case 'd':
          dimensions = atoi(optarg);
          break;
        case 's':
          if (!topology_set) {
            topology = 's';
            topology_set = true;
            num_simplices = atoi(optarg);
          } else {
            error_flag = true;
          }
          break;
        case 't':
          if (!topology_set) {
            topology = 't';
            topology_set= true;
            num_simplices = atoi(optarg);
          } else {
            error_flag = true;
          }
          break;
        case '?':
          if (optopt == 's' || optopt =='t' || optopt == 'd') {
            std::cout << "Option -" << optopt << " requires an argument."
            << std::endl;
          } else {
            std::cerr << "Uknown option -" << optopt << std::endl;
          }
        print_error(argv[0]);
        return 1;
      default:
        abort();
      }

  if (error_flag || num_simplices == 0 || dimensions > 3) {
    print_error(argv[0]);
    exit(2);
  }

  /// Display job parameters
  std::cout << "Number of dimensions = " << dimensions << std::endl;
  std::cout << "Number of simplices = " << num_simplices << std::endl;
  std::cout << "Geometry = " << (topology == 's'
    ? "spherical" : "toroidal") << std::endl;
  std::cout << "User = " << getEnvVar("USER") << std::endl;
  std::cout << "Hostname = " << hostname() << std::endl;

  /// Initialize both simplicial complex types
  Delaunay Sphere3;
  PDT Torus3;

  switch (topology) {
    case 's':
      make_S3_simplicial_complex(&Sphere3, num_simplices);
      t.stop();
      print_results(&Sphere3, &t);
      write_file(Sphere3, topology, dimensions, num_simplices);
      break;
    case 't':
      make_T3_simplicial_complex(&Torus3, num_simplices);
      t.stop();
      print_results(&Torus3, &t);
      write_file(Torus3, topology, dimensions, num_simplices);
      break;
    default:
      print_error(argv[0]);
      return 1;
  }
  return 0;
}

File: cdt.py
#!/usr/bin/env python

"""Causal Dynamical Triangulations in Python using CGAL.

Copyright (c) 2014 Adam Getchell

A program that generates d-dimensional triangulated spacetimes
with a defined causal structure and evolves them according
to the Metropolis algorithm.

This is for quick and easy development in Python
before writing in C++ for performance.

Uses https://code.google.com/p/cgal-bindings/

Usage:
  cdt.py  ( --spherical | --toroidal | --periodic )
          ( -n SIMPLICES)
          ( -t TIMESLICES)
          [ -d DIMENSIONS]
  cdt.py --version
  cdt.py --help

Examples:
  cdt.py --spherical -n 5000 -t 256
  cdt-py --periodic -n 5000 -t 256 -d 3

Options:
  -h --help         Show this screen.
  -v --version      Show version.
  --spherical       Spherical topology
  --toroidal        Toroidal topology
  --periodic        Toroidal topology
  -n SIMPLICES      Total number of simplices
  -t TIMESLICES     Number of timeslices
  -d DIMENSIONS     Dimensionality of triangulation [default: 3]

"""

from docopt import docopt
import os
import socket
import utilities
import sys
import spherical_3_triangulations as s3
from CGAL.CGAL_Triangulation_3 import Delaunay_triangulation_3

if __name__ == '__main__':
  arguments = docopt(__doc__, version='0.1')

  print "Number of dimensions = ", arguments['-d']
  print "Number of simplices = ", arguments['-n']
  print "Number of timeslices = ", arguments['-t']
  print "Topology is ", "spherical" if arguments['--spherical'] else "periodic"
  print "User = ", os.getlogin()
  print "Hostname = ", socket.gethostname()

  # To debug docopt uncomment the following line
  # print arguments

  dimensions = int(arguments['-d'])
  simplices = int(arguments['-n'])
  timeslices = int(arguments['-t'])

  spherical = True if arguments['--spherical'] else False
  filename = utilities.generate_filename(spherical,
                                          str(dimensions),
                                          str(simplices),
                                          str(timeslices))

  if dimensions != 3:
    sys.exit("Only 3D triangulations currently supported")
  elif arguments['--spherical']:
    S = s3.make_S3_triangulations(dimensions, simplices, timeslices)
  else:
    print "Call make_T3_triangulations()"

  S.write_to_file(filename, 14)
  print "Results written to ", filename

  T1 = Delaunay_triangulation_3()
  T1.read_from_file(filename)

  assert T1.is_valid()
  assert T1.number_of_vertices() == S.number_of_vertices()
  assert T1.number_of_cells() == S.number_of_cells()
  print "Results verified"
  print "Final triangulation has "
  utilities.print_results(T1)
